src/mpi/pt2pt/recv.c

```diff
@@ -145,6 +145,25 @@ int MPI_Recv(void *buf, int count, MPI_Datatype datatype, int source, int tag,
         goto fn_exit;
     }
 
+    if (unlikely(MPIR_CVAR_ENABLE_FT &&
+                 !MPIR_Request_is_complete(request_ptr) &&
+                 MPID_Request_is_anysource(request_ptr) &&
+                 !MPID_Comm_AS_enabled(request_ptr->comm))) {
+        /* --BEGIN ERROR HANDLING-- */
+        MPID_Cancel_recv(request_ptr);
+        MPIR_STATUS_SET_CANCEL_BIT(request_ptr->status, FALSE);
+        MPIR_ERR_SET(request_ptr->status.MPI_ERROR, MPIX_ERR_PROC_FAILED, "**proc_failed");
+        mpi_errno = request_ptr->status.MPI_ERROR;
+        goto fn_fail;
+        /* --END ERROR HANDLING-- */
+    }
+
+    if (MPIR_ThreadInfo.isThreaded) {
```

Comment:
I'd prefer if there was a way to always call `MPID_Wait` instead of have this branch here. However, this is kind of a special case where we know what type of request we are waiting on (i.e. not a Generalized Request), so there is potential benefit to having the specialized wait loop in the non-multithreaded case. Not sure the best way to proceed here...